keps/sig-storage/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,590 @@
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Volume reclaim policy for the StatefulSet created PVCs
+
+When a statefulset spec has a `VolumeClaimTemplate`, PVCs are dynamically created 
+using a static naming scheme. A new field named `PersistentVolumeClaimReclaimPolicy` of the 
+type `StatefulSetPersistentVolumeClaimReclaimPolicy` will be added to the StatefulSet. This 
+field will represent the user indication on whether the associated PVCs can be automatically 
+deleted or not. The default policy would be `Retain`. 
+
+If `PersistentVolumeClaimReclaimPolicy` is set to `RemoveOnScaledown`, Pod is set as the owner of the PVCs created
+from the `VolumeClaimTemplates` just before the scale down is performed by the statefulset controller. 
+When a Pod is deleted, the PVC owned by the Pod is also deleted. When `RemoveOnScaledown` 
+policy is set and the Statefulset gets deleted the PVCs also will get deleted 
+(similar to `RemoveonStatefulSetDeletion` policy).
+
+Current scaleset controller implementation ensures that the manually deleted pods are restored 
+before the scale down logic is run. This combined with the fact that the owner references are set 
+only before the scale down will ensure that manual deletions do not automatically delete the PVCs 
+in question.
+
+During scale-up, if a PVC has an OwnerRef that does not match the Pod, it 
+potentially indicates that the PVC is referred by the deleted Pod and is in the process of 
+getting deleted. Controller will exit the current reconcile loop and attempt to reconcile in the 
+next iteration. This avoids a race with PVC deletion.
+
+When `PersistentVolumeClaimReclaimPolicy` is set to `RemoveOnStatefulSetDeletion` the owner reference in 
+PVC points to the StatefulSet. When a scale up or down occurs, the PVC would remain the same. 
+PVCs previously in use before scale down will be used again when the scale up occurs. The PVC deletion 
+should happen only after the Pod gets deleted. Since the Pod ownership has `blockOwnerDeletion` set to 
+`true` pods will get deleted before the StatefulSet is deleted. The `blockOwnerDeletion` for PVCs will 
+be set to `false` which ensures that PVC deletion happens only after the StatefulSet is deleted. This 
+chain of ownership ensures that Pod deletion occurs before the PVCs are deleted.
```

Comment:
What would happen if I delete a `RemoveOnStatefulSetDeletion` StatefulSet with orphan policy? Will the PVCs be gone while the Pods stay?

Reply:
By the orphan policy do you mean kubectl delete --cascade=false? 

In that case I think nothing should be deleted other than the statefulset resource, the semantics for cascade seem clear. That would mean removing any ownership IIUC.

At any rate, even if the PVCs were deleted, the protection controller would prevent them from being finalized until the pods were deleted. But it might be unexpected to have the PVCs disappear when the pods are eventually deleted.